corehq/apps/es/README.rst

```diff
@@ -1,6 +1,174 @@
+framework cannot automatically determine what operations need to be included in
+a migration, or even when a new migration is required. This is why creating
+these migrations is a separate command and not integrated into the default
+``makemigrations`` command.
+
+To create a new Elasticsearch index migration, use the
+``make_elastic_migration`` management command and provide details for the
+required migration operations via any combination of the ``-c/--create``,
+``-u/--update`` and/or ``-d/--delete`` command line options.
+
+Similar to Django model migrations, this management command uses the index
+metadata (mappings, analysis, etc) from the existing Elasticsearch code, so it
+is important that this command is executed *after* making changes to index
+metadata. To provide an example, consider a hypothetical scenario where the
+following index changes are needed:
+
+- create a new ``users`` index
+- update the mapping on the existing ``groups`` index to add a new property
+  named ``pending_users``
+- delete the existing index named ``groups-sandbox``
+
+After the new property has been added to the ``groups`` index mapping in code,
+the following management command would create a migration file (e.g.
+``corehq/apps/es/migrations/0003_groups_pending_users.py``) for the necessary
+operations:
+
+.. code-block:: shell
+
+    ./manage.py make_elastic_migration --name groups_pending_users -c users -u groups:pending_users -d groups-sandbox
+
+
+.. _updating-elastic-index-mappings:
+
+Updating Elastic Index Mappings
+'''''''''''''''''''''''''''''''
+
+Prior to the ``UpdateIndexMapping`` migration operation implementation, Elastic
+mappings were always applied "in full" any time a mapping change was needed.
+That is: the entire mapping (from code) was applied to the existing index via
+the `Put Mapping`_ API. This technique had some pros and cons:
+
+- **Pro**: the mapping update logic in code was simple because it did not have
+  to worry about which *existing* mapping properties are persistent (persist on
+  the index even if omitted in a PUT request payload) and which ones are
+  volatile (effectively "unset" if omitted in a PUT request payload).
+- **Con**: it requires that *all* mapping properties are explicitly set on every
+  mapping update, making mapping updates impossible if the existing index
+  mapping in Elasticsearch has diverged from the mapping in code.
+
+Because CommCare HQ Elastic mappings have been able to drift between
+environments, it is no longer possible to update some index mappings using the
+current technique. On some indexes, the live index mappings have sufficiently
+diverged that there is no common, "full mapping definition" that can be applied
+on all environments. This means that in order to push mapping changes to all
+environments, new mapping update logic is needed which is capable of updating
+individual properties on an Elastic index mapping while leaving other (existing)
+properties unchanged.
```

Comment:
Seems like this paragraph will become outdated as soon as we adopt the new migrations framework. What do you think of rephrasing it to be a perspective on past practice rather than "current" state?

I'm thinking specifically of this phrase:

> it is no longer possible to update some index mappings using the current technique

Might be as simple as using the word "historical" rather than "current"?

Reply:
See fe7b237e490